## Data Overview
- VIN (1-10) - The 1st 10 characters of each vehicle's Vehicle Identification Number (VIN).
- County- The county in which the registered owner resides.
- City - The city in which the registered owner resides.
- State- The state in which the registered owner resides.
- Postal Code - The 5-digit postal code in which the registered owner resides.
- Model Year - The model year of the vehicle is determined by decoding the Vehicle Identification Number (VIN).
- Make- The manufacturer of the vehicle, determined by decoding the Vehicle Identification Number (VIN).
- Model- The model of the vehicle is determined by decoding the Vehicle Identification Number (VIN).
- Electric Vehicle Type - This distinguishes the vehicle as all-electric or a plug-in hybrid.
- Clean Alternative Fuel Vehicle (CAFV) Eligibility - This categorizes vehicles as Clean Alternative Fuel Vehicles (CAFVs) based on the fuel requirement and electric-only range requirement.
- Electric Range - Describes how far a vehicle can travel purely on its electric charge.
- Base MSRP - This is the lowest Manufacturer's Suggested Retail Price (MSRP).
- Legislative District - The specific section of Washington State that the vehicle's owner resides in, as represented in the state legislature.
- DOL Vehicle ID - Unique number assigned to each vehicle by the Department of Licensing for identification purposes.
- Vehicle Location - The center of the ZIP Code for the registered vehicle.
- Electric Utility - This is the electric power retail service territory serving the address of the registered vehicle.
- 2020 Census Tract - A census tract is a small, relatively permanent statistical subdivision of a county delineated by a local committee of census data users for the purpose of presenting census data in 2020.

In [1]:
# Importing all the required libraries.
import numpy as np
import pandas as pd

import seaborn as sns
import plotly.express as px

In [2]:
# Loading the EV(Electric Vehicle) dataset.
ev_data = pd.read_csv(r"D:\Data\EVdataset.csv")

In [3]:
# Checking the head of the dataframe.
ev_data.head(10)

VIN (1-10)     County       City State  Postal Code  Model Year       Make  \
0  JTMEB3FV6N     Monroe   Key West    FL        33040        2022     TOYOTA   
1  1G1RD6E45D      Clark   Laughlin    NV        89029        2013  CHEVROLET   
2  JN1AZ0CP8B     Yakima     Yakima    WA        98901        2011     NISSAN   
3  1G1FW6S08H     Skagit   Concrete    WA        98237        2017  CHEVROLET   
4  3FA6P0SU1K  Snohomish    Everett    WA        98201        2019       FORD   
5  5YJ3E1EB5J  Snohomish    Bothell    WA        98021        2018      TESLA   
6  1N4AZ0CP4D  Snohomish    Everett    WA        98203        2013     NISSAN   
7  1N4AZ0CP0D  Snohomish   Mukilteo    WA        98275        2013     NISSAN   
8  1N4BZ0CP4G     Island    Clinton    WA        98236        2016     NISSAN   
9  KNDJP3AE2G     Skagit  Anacortes    WA        98221        2016        KIA   

        Model                   Electric Vehicle Type  \
0  RAV4 PRIME  Plug-in Hybrid Electric Vehicle (PHEV)   
1        VOLT  Plug-in Hybrid Electric Vehicle (PHEV)   
2        LEAF          Battery Electric Vehicle (BEV)   
3     BOLT EV          Battery Electric Vehicle (BEV)   
4      FUSION  Plug-in Hybrid Electric Vehicle (PHEV)   
5     MODEL 3          Battery Electric Vehicle (BEV)   
6        LEAF          Battery Electric Vehicle (BEV)   
7        LEAF          Battery Electric Vehicle (BEV)   
8        LEAF          Battery Electric Vehicle (BEV)   
9        SOUL          Battery Electric Vehicle (BEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0           Clean Alternative Fuel Vehicle Eligible              42   
1           Clean Alternative Fuel Vehicle Eligible              38   
2           Clean Alternative Fuel Vehicle Eligible              73   
3           Clean Alternative Fuel Vehicle Eligible             238   
4             Not eligible due to low battery range              26   
5           Clean Alternative Fuel Vehicle Eligible             215   
6           Clean Alternative Fuel Vehicle Eligible              75   
7           Clean Alternative Fuel Vehicle Eligible              75   
8           Clean Alternative Fuel Vehicle Eligible              84   
9           Clean Alternative Fuel Vehicle Eligible              93   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0          0                   NaN       198968248   
1          0                   NaN         5204412   
2          0                  15.0       218972519   
3          0                  39.0       186750406   
4          0                  38.0         2006714   
5          0                   1.0       475635324   
6          0                  38.0       253546023   
7          0                  21.0       176817364   
8          0                  10.0       475900379   
9      31950                  40.0       100489049   

              Vehicle Location        Electric Utility  2020 Census Tract  
0    POINT (-81.80023 24.5545)                     NaN        12087972100  
1  POINT (-114.57245 35.16815)                     NaN        32003005702  
2  POINT (-120.50721 46.60448)              PACIFICORP        53077001602  
3   POINT (-121.7515 48.53892)  PUGET SOUND ENERGY INC        53057951101  
4  POINT (-122.20596 47.97659)  PUGET SOUND ENERGY INC        53061041500  
5   POINT (-122.18384 47.8031)  PUGET SOUND ENERGY INC        53061051916  
6  POINT (-122.23019 47.94949)  PUGET SOUND ENERGY INC        53061040900  
7  POINT (-122.29196 47.89908)  PUGET SOUND ENERGY INC        53061042001  
8   POINT (-122.35803 47.9796)  PUGET SOUND ENERGY INC        53029972100  
9  POINT (-122.61214 48.51748)  PUGET SOUND ENERGY INC        53057940301

In [4]:
# Checking the shape of the dataframe.
ev_data.shape

(112634, 17)

In [5]:
# Checking the Info() of the dataframe.
ev_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112634 entries, 0 to 112633
Data columns (total 17 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         112634 non-null  object 
 1   County                                             112634 non-null  object 
 2   City                                               112634 non-null  object 
 3   State                                              112634 non-null  object 
 4   Postal Code                                        112634 non-null  int64  
 5   Model Year                                         112634 non-null  int64  
 6   Make                                               112634 non-null  object 
 7   Model                                              112614 non-null  object 
 8   Electric Vehicle Type                              112634 non-null  object

In [6]:
# checking the descritive statistics of the data.
ev_data.describe()

Postal Code     Model Year  Electric Range      Base MSRP  \
count  112634.000000  112634.000000   112634.000000  112634.000000   
mean    98156.226850    2019.003365       87.812987    1793.439681   
std      2648.733064       2.892364      102.334216   10783.753486   
min      1730.000000    1997.000000        0.000000       0.000000   
25%     98052.000000    2017.000000        0.000000       0.000000   
50%     98119.000000    2020.000000       32.000000       0.000000   
75%     98370.000000    2022.000000      208.000000       0.000000   
max     99701.000000    2023.000000      337.000000  845000.000000   

       Legislative District  DOL Vehicle ID  2020 Census Tract  
count         112348.000000    1.126340e+05       1.126340e+05  
mean              29.805604    1.994567e+08       5.296650e+10  
std               14.700545    9.398427e+07       1.699104e+09  
min                1.000000    4.777000e+03       1.101001e+09  
25%               18.000000    1.484142e+08       5.303301e+10  
50%               34.000000    1.923896e+08       5.303303e+10  
75%               43.000000    2.191899e+08       5.305307e+10  
max               49.000000    4.792548e+08       5.603300e+10

## *`Exploratory Data Analysis`*

### *`Checking Duplicates`*

In [7]:
# Checking if any duplicates are present in the dataframe.
ev_data.duplicated().value_counts()

False    112634
dtype: int64

### *`Checking Null values`*

In [8]:
# Checking the null values 
ev_data.isnull().sum()

VIN (1-10)                                             0
County                                                 0
City                                                   0
State                                                  0
Postal Code                                            0
Model Year                                             0
Make                                                   0
Model                                                 20
Electric Vehicle Type                                  0
Clean Alternative Fuel Vehicle (CAFV) Eligibility      0
Electric Range                                         0
Base MSRP                                              0
Legislative District                                 286
DOL Vehicle ID                                         0
Vehicle Location                                      24
Electric Utility                                     443
2020 Census Tract                                      0
dtype: int64

In [9]:
# Checking the percentage of missing or null values.
np.round((ev_data.isnull().sum()/len(ev_data))*100,2).astype('str')+'%'

VIN (1-10)                                            0.0%
County                                                0.0%
City                                                  0.0%
State                                                 0.0%
Postal Code                                           0.0%
Model Year                                            0.0%
Make                                                  0.0%
Model                                                0.02%
Electric Vehicle Type                                 0.0%
Clean Alternative Fuel Vehicle (CAFV) Eligibility     0.0%
Electric Range                                        0.0%
Base MSRP                                             0.0%
Legislative District                                 0.25%
DOL Vehicle ID                                        0.0%
Vehicle Location                                     0.02%
Electric Utility                                     0.39%
2020 Census Tract                                     0.

### 🌟`Important Observations` :
- There are total 0.68% percentage of null values are present in the data.There are two ways to treat null values.They are 
1. Dropping missing or Null values
- remove rows having missing values if overall less than 5% of rows are missing in the DataFrame.
- remove the whole column containing missing values - go for this approach if more than 50% of the column data is missing.
2. Imputation methods for imputing the null values like Simple Imputer and Knn Imputer.

In [10]:
# View the missing or Null values
ev_data[ev_data.Model.isna()]

VIN (1-10)     County          City State  Postal Code  Model Year  \
13874   YV4ED3GM2P       King       Seattle    WA        98115        2023   
30517   YV4ED3UL3P       King       Seattle    WA        98115        2023   
31936   YV4ED3GM4P    Clallam        Sequim    WA        98382        2023   
37517   YV4ED3UW2P  Snohomish       Edmonds    WA        98026        2023   
58071   YV4ED3UM4P       King        Renton    WA        98058        2023   
61626   YV4ED3GM5P     Pierce        Tacoma    WA        98465        2023   
63240   YV4ED3GMXP       King       Redmond    WA        98052        2023   
63380   YV4ED3GM7P       King       Seattle    WA        98122        2023   
63462   YV4ED3UW4P       King     Newcastle    WA        98059        2023   
78472   YV4ED3UM1P       King     Fall City    WA        98024        2023   
81302   YV4ED3UM5P       King       Redmond    WA        98052        2023   
84142   YV4ED3UM2P       King    North Bend    WA        98045        2023   
86960   YV4ED3UM9P       King     Sammamish    WA        98075        2023   
88687   YV4ED3GM5P       King  Maple Valley    WA        98038        2023   
89882   YV4ED3UM5P       King      Bellevue    WA        98006        2023   
93197   YV4ED3GM8P  Snohomish       Bothell    WA        98021        2023   
103099  YV4ED3UW6P     Pierce        Milton    WA        98354        2023   
103394  YV4ED3GM5P       King       Seattle    WA        98133        2023   
108116  YV4ED3GL1P       King       Seattle    WA        98104        2023   
112622  YV4ED3GM0P       King     Covington    WA        98042        2023   

         Make Model           Electric Vehicle Type  \
13874   VOLVO   NaN  Battery Electric Vehicle (BEV)   
30517   VOLVO   NaN  Battery Electric Vehicle (BEV)   
31936   VOLVO   NaN  Battery Electric Vehicle (BEV)   
37517   VOLVO   NaN  Battery Electric Vehicle (BEV)   
58071   VOLVO   NaN  Battery Electric Vehicle (BEV)   
61626   VOLVO   NaN  Battery Electric Vehicle (BEV)   
63240   VOLVO   NaN  Battery Electric Vehicle (BEV)   
63380   VOLVO   NaN  Battery Electric Vehicle (BEV)   
63462   VOLVO   NaN  Battery Electric Vehicle (BEV)   
78472   VOLVO   NaN  Battery Electric Vehicle (BEV)   
81302   VOLVO   NaN  Battery Electric Vehicle (BEV)   
84142   VOLVO   NaN  Battery Electric Vehicle (BEV)   
86960   VOLVO   NaN  Battery Electric Vehicle (BEV)   
88687   VOLVO   NaN  Battery Electric Vehicle (BEV)   
89882   VOLVO   NaN  Battery Electric Vehicle (BEV)   
93197   VOLVO   NaN  Battery Electric Vehicle (BEV)   
103099  VOLVO   NaN  Battery Electric Vehicle (BEV)   
103394  VOLVO   NaN  Battery Electric Vehicle (BEV)   
108116  VOLVO   NaN  Battery Electric Vehicle (BEV)   
112622  VOLVO   NaN  Battery Electric Vehicle (BEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
13874   Eligibility unknown as battery range has not b...               0   
30517   Eligibility unknown as battery range has not b...               0   
31936   Eligibility unknown as battery range has not b...               0   
37517   Eligibility unknown as battery range has not b...               0   
58071   Eligibility unknown as battery range has not b...               0   
61626   Eligibility unknown as battery range has not b...               0   
63240   Eligibility unknown as battery range has not b...               0   
63380   Eligibility unknown as battery range has not b...               0   
63462   Eligibility unknown as battery range has not b...               0   
78472   Eligibility unknown as battery range has not b...               0   
81302   Eligibility unknown as battery range has not b...               0   
84142   Eligibility unknown as battery range has not b...               0   
86960   Eligibility unknown as battery range has not b...               0   
88687   Eligibility unknown as battery range has not b...               0   
89882   Eligibility unknown as battery range has not b... 

### 🌟`Important Observations` :
- Byseeing the data of Null or Missing values in the where make is Volvo and model year is 2023 there we have most of the null values model and might be the reson for the null values.

In [11]:
ev_data[ev_data['Legislative District'].isna()]

VIN (1-10)        County               City State  Postal Code  \
0       JTMEB3FV6N        Monroe           Key West    FL        33040   
1       1G1RD6E45D         Clark           Laughlin    NV        89029   
12      3C3CFFGE3G     St. Clair          Mascoutah    IL        62258   
19      5YJXCAE28G      Saratoga  Greenfield Center    NY        12833   
41      5YJSA1E22G  Newport News       Newport News    VA        23602   
...            ...           ...                ...   ...          ...   
112153  1G1RB6E46F       Carroll       North Conway    NH         3860   
112301  5YJ3E1EB9M    Dorchester        Summerville    SC        29483   
112394  5UXKT0C36H   Leavenworth            Lansing    KS        66043   
112541  JA4J24A50J      Williams          Williston    ND        58802   
112603  7FCTGAAL7N      Kootenai             Worley    ID        83876   

        Model Year        Make       Model  \
0             2022      TOYOTA  RAV4 PRIME   
1             2013   CHEVROLET        VOLT   
12            2016        FIAT         500   
19            2016       TESLA     MODEL X   
41            2016       TESLA     MODEL S   
...            ...         ...         ...   
112153        2015   CHEVROLET        VOLT   
112301        2021       TESLA     MODEL 3   
112394        2017         BMW          X5   
112541        2018  MITSUBISHI   OUTLANDER   
112603        2022      RIVIAN         R1T   

                         Electric Vehicle Type  \
0       Plug-in Hybrid Electric Vehicle (PHEV)   
1       Plug-in Hybrid Electric Vehicle (PHEV)   
12              Battery Electric Vehicle (BEV)   
19              Battery Electric Vehicle (BEV)   
41              Battery Electric Vehicle (BEV)   
...                                        ...   
112153  Plug-in Hybrid Electric Vehicle (PHEV)   
112301          Battery Electric Vehicle (BEV)   
112394  Plug-in Hybrid Electric Vehicle (PHEV)   
112541  Plug-in Hybrid Electric Vehicle (PHEV)   
112603          Battery Electric Vehicle (BEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0                 Clean Alternative Fuel Vehicle Eligible              42   
1                 Clean Alternative Fuel Vehicle Eligible              38   
12                Clean Alternative Fuel Vehicle Eligible              84   
19                Clean Alternative Fuel Vehicle Eligible             200   
41                Clean Alternative Fuel Vehicle Eligible             210   
...                                                   ...             ...   
112153            Clean Alternative Fuel Vehicle Eligible              38   
112301  Eligibility unknown as battery range has not b...               0   
112394              Not eligible due to low battery range              14   
112541              Not eligible due to low battery range              22   
112603  Eligibility unknown as battery range has not b...               0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
0               0                   NaN       198968248   
1               0                   NaN         5204412   
12              0                   NaN       153786167   
19              0                   NaN       218050878   
41              0                   NaN       111593331   
...           ...                   ...             ...   
112153          0                   NaN       177816061   
112301          0                   NaN       179604183   
112394          0                   NaN       122897484   
112541          0                   NaN         2592005   
112603          0                   NaN       211894693   

                   Vehicle Location Electric Utility  2020 Census Tract  
0         POINT (-81.80023 24.5545)              NaN        12087972100  
1       POINT (-114.57245 35.16815)              NaN        32003005702  
12       POINT (-89.79939 38.49028)              NaN        17163504356  
19        POINT (-73.84643 43.1284)

### 🌟`Important Observations` :
- Byseeing the data of Null or Missing values Imputation methods is the best option for imputing the null values using Simple Imputer because we donot lose the data.

In [12]:
ev_data.Model.value_counts()

MODEL 3        23135
MODEL Y        17142
LEAF           12880
MODEL S         7377
BOLT EV         4910
               ...  
745LE              2
S-10 PICKUP        1
SOLTERRA           1
918                1
FLYING SPUR        1
Name: Model, Length: 114, dtype: int64

In [13]:
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(strategy='most_frequent')
imputer.fit(ev_data[['Model','Legislative District','Vehicle Location','Electric Utility']])
ev_data['Model'] = imputer.fit_transform(ev_data[['Model']])
ev_data['Legislative District'] = imputer.fit_transform(ev_data[['Legislative District']])
ev_data['Vehicle Location'] = imputer.fit_transform(ev_data[['Vehicle Location']])
ev_data['Electric Utility'] = imputer.fit_transform(ev_data[['Electric Utility']])

In [14]:
ev_data.isna().sum()

VIN (1-10)                                           0
County                                               0
City                                                 0
State                                                0
Postal Code                                          0
Model Year                                           0
Make                                                 0
Model                                                0
Electric Vehicle Type                                0
Clean Alternative Fuel Vehicle (CAFV) Eligibility    0
Electric Range                                       0
Base MSRP                                            0
Legislative District                                 0
DOL Vehicle ID                                       0
Vehicle Location                                     0
Electric Utility                                     0
2020 Census Tract                                    0
dtype: int64

In [15]:
ev_data.iloc[13874]

VIN (1-10)                                                                                  YV4ED3GM2P
County                                                                                            King
City                                                                                           Seattle
State                                                                                               WA
Postal Code                                                                                      98115
Model Year                                                                                        2023
Make                                                                                             VOLVO
Model                                                                                          MODEL 3
Electric Vehicle Type                                                   Battery Electric Vehicle (BEV)
Clean Alternative Fuel Vehicle (CAFV) Eligibility    Eligibility unknown 

In [16]:
ev_data.Model.value_counts()

MODEL 3        23155
MODEL Y        17142
LEAF           12880
MODEL S         7377
BOLT EV         4910
               ...  
745LE              2
S-10 PICKUP        1
SOLTERRA           1
918                1
FLYING SPUR        1
Name: Model, Length: 114, dtype: int64

### *`Checking If any corrupted values are present`*

In [17]:
# Identifying the corrupted values by seeing info() of the data.
ev_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112634 entries, 0 to 112633
Data columns (total 17 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         112634 non-null  object 
 1   County                                             112634 non-null  object 
 2   City                                               112634 non-null  object 
 3   State                                              112634 non-null  object 
 4   Postal Code                                        112634 non-null  int64  
 5   Model Year                                         112634 non-null  int64  
 6   Make                                               112634 non-null  object 
 7   Model                                              112634 non-null  object 
 8   Electric Vehicle Type                              112634 non-null  object

### *`Checking or Identifying the outliers`*

In [18]:
ev_data.skew(numeric_only=True)

Postal Code            -27.955285
Model Year              -0.817579
Electric Range           0.817103
Base MSRP               10.100497
Legislative District    -0.546495
DOL Vehicle ID           1.149258
2020 Census Tract      -25.013927
dtype: float64

### 🌟`Important Observations` :
- Legislative District skewness is -0.54. so, the data is fairly symetrical.
- Model Year, Electric Range, DOL Vehicle ID are moderately skewed.
- Postal Code and 2020 Census Tract are highly skewed.

In [ ]:
px.box(data_frame=ev_data, x='Model Year')

In [ ]:
px.box(ev_data, x='Electric Range')

In [ ]:
px.box(ev_data, x='Base MSRP')

In [22]:
ev_data['Base MSRP'].value_counts()

0         109122
69900       1497
31950        407
52900        214
32250        160
54950        135
59900        134
39995        119
36900        100
44100         97
64950         83
33950         78
45600         76
52650         67
34995         58
36800         50
55700         47
53400         28
110950        24
98950         23
102000        20
90700         19
81100         19
75095         16
184400        12
43700         10
109000         7
89100          7
91250          4
845000         1
Name: Base MSRP, dtype: int64

### 🌟`Important Observations` :
- There are more most of the zeros in the Base MSRP column so, all the other values are falling as outliers because of that.

In [ ]:
px.box(ev_data, x='Legislative District')

In [ ]:
px.box(ev_data, x='DOL Vehicle ID')

In [ ]:
px.box(ev_data, x='2020 Census Tract')

### Task1 - Apply Exploratory Data Analysis(Univariate and Bivariate) using plotly.express library.

In [ ]:
px.histogram(ev_data, x="Model", title="Frequency Distribution of Models in Electric Vehicle data")

### 🌟`Important Observations` :
- Model 3 is the Most Frequent Model in the data with the count of 23.155k and Model Y is the second most frequent Model in the data with the count of 17.142k.
- Model Leaf is also a most frequent Model and with the count of 12.88K.

In [ ]:
px.histogram(ev_data,x='Model Year', title="Year wise Frequency Distributions")

### 🌟`Important Observations` :
- Production of Electric vehicles is increasing year by year and sightly we can observe the there is also a rise fall from 2018 to 2019.
- In 2022 there is highest production of Electric vehicles and reached upto 26,53k vechiles.

In [ ]:
px.pie(names=ev_data.Make.value_counts().index, values=ev_data.Make.value_counts().values,
       width=1000,height=700, title="Distribution of make(manufacturer of the vehicle)")

### 🌟`Important Observations` :
- Most of the Electric vehicles was manufactured by Tesla with 46.2%(52,078 Vechiles) and followed by the Nissan Company has produced 11.4%(12,880 Vechiles).
- Azure Dynamics has manufactured lowest no of Electric Vechiles that are only 7 Vechiles.

### Does an EV have a battery? 
- Yes Electric Vehicles has the battery in it and operates only on electricity. those are battery-powered electric vehicle (BEV).
- EV industry is Evolving around the Battery Vehicles to achieve Zero or No emissions.

### Is hybrid a type of EV?
- Yes by seeeing the data we can say that they are hybrid electric Vehicles they are plug- in Hybrid electric vehicle.

In [ ]:
px.pie(names=ev_data['Electric Vehicle Type'].value_counts().index, 
       values=ev_data['Electric Vehicle Type'].value_counts().values, title="Distribution of Electric Vehicle Type")

### 🌟`Important Observations` :
- Most of the Electric vehicles are Battery Eletric Vehicle(BEV) exactly 76.4%(86,044) and ew can say that the EV industry is completely moving to the battery side.

In [ ]:
px.pie(names=ev_data['Clean Alternative Fuel Vehicle (CAFV) Eligibility'].value_counts().index,
       values=ev_data['Clean Alternative Fuel Vehicle (CAFV) Eligibility'].value_counts().values,
      title="Distribution of Clean Alternative Fuel Vehicle (CAFV) Eligibility")

### 🌟`Important Observations` :
- Count of CAFV eligibility(Clean Alternative Fuel Vehicle Eligible) and Most of the Electric Vehicles are Eligible with the 52.1%.
- Majority Eligibility Category is **`CAFV(Clean Alternative Fuel Vehicle)`** and Minority Eligibility Categories **`Not eligible due to low battery range`**.
- There is one more category not eligible because of the battery they are **`Eligibility unknown as battery range has not been researched`** with 34.8%.

In [31]:
ev_data['Clean Alternative Fuel Vehicle (CAFV) Eligibility'].value_counts()

Clean Alternative Fuel Vehicle Eligible                         58639
Eligibility unknown as battery range has not been researched    39236
Not eligible due to low battery range                           14759
Name: Clean Alternative Fuel Vehicle (CAFV) Eligibility, dtype: int64

In [ ]:
px.scatter(ev_data, x='Make', y='Model Year', title=' Year wise manufacturing of Electric Vehicles by the companies')

### 🌟`Important Observations` :
- By observing the data we can say that **`Chevrolet`** is the first manufacturing company started the jounrey of electric vehicle in 1997.
- after that Toyota, ford and Tesla etc.,. are started there jounrey in the EV Industry.
- In 2023 most of the manufacture company's are buliding the electric vechile and producing the more electric vechile for the country.

In [ ]:
px.scatter(ev_data, x='Make', y='Electric Range', title="Electric Vehicle Range by Make")

### 🌟`Important Observations` :
- These shows the distribution of electric ranges for different vehicle makes(manufatures) and also Variability of Electric Range of Across Makes.
- **`Tesla`** Manufacture with Highest Electric Range compared other makes or manufactures with the electric range of 337.
- There are many manufactures(Make) with lowest Electric Range with 0.

In [ ]:
px.scatter(ev_data, x='Make', y='Electric Range', hover_data='Model')

### 🌟`Important Observations` :
- In **`Tesla`** Manufacture Model S is with Highest Electric Range compared other makes or manufactures with the electric range of 337.

### Task2 - Create a Choropleth to display the number of EV vehicles based on location.

In [35]:
ev_count_df = ev_data.groupby('State').size().reset_index(name='EV Count')

In [36]:
ev_count_df.head()

State  EV Count
0    AK         1
1    AL         1
2    AR         4
3    AZ         6
4    CA        76

In [ ]:
fig = px.choropleth(
    ev_count_df,
    locations='State',  # Column in the DataFrame containing the location names
    locationmode='USA-states',  # Use 'country names' if using country-level data
    color='EV Count',  # Column in the DataFrame containing the EV vehicle count
    hover_name='State',  # Hover text that appears when hovering over a location
    title='Number of EV Vehicles by Location'
)

fig.show()

### Task3 - Create a Racing Bar Plot to display the animation of EV Make and its count each year.

In [39]:
# Group by model year and make, and calculate the count for each group.
df_grouped = ev_data.groupby(['Model Year', 'Make']).size().reset_index(name='Count')

In [40]:
df_grouped

Model Year        Make  Count
0          1997   CHEVROLET      1
1          1998        FORD      1
2          1999        FORD      3
3          2000        FORD     10
4          2002      TOYOTA      2
..          ...         ...    ...
204        2023    POLESTAR     88
205        2023      SUBARU      1
206        2023       TESLA    890
207        2023  VOLKSWAGEN     69
208        2023       VOLVO     21

[209 rows x 3 columns]

In [41]:
import bar_chart_race as bcr

In [ ]:
bcr.bar_chart_race(df=df_grouped,
                   orientation='h',
                   sort='desc',  # sort the bar for each period
                   n_bars=10, # number of bars to display in each frame
                   steps_per_period=45, # smoothness of the animation
                   period_length=3000, # time period in ms for each row
                   title={'label': 'Year wise sales of each Makers from last decade','size': 52, 
                          'weight': 'bold', 'pad': 40 })

In [ ]:
fig = px.bar(df_grouped,
             x="Make",
             y="Count",
             animation_frame="Model Year",
             range_y=[0, df_grouped['Count'].max() + 50],
             title="EV Make Count Each Year",
             labels={'Make': 'EV Make', 'Count': 'Number of EVs'},
             template="plotly_dark"
            )

# Show the figure
fig.show()